modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -417,14 +428,42 @@ void onBulkResponse(BulkResponse response, Runnable onSuccess) {
         }
     }
 
-    void notifyDone(long thisBatchStartTimeNS, ScrollableHitSource.AsyncResponse asyncResponse, int batchSize) {
+    void notifyDone(long thisBatchStartTimeNS,
+                    ScrollableHitSource.AsyncResponse asyncResponse,
+                    int batchSize,
+                    List<? extends ScrollableHitSource.Hit> scrollRemainingHits) {
         if (task.isCancelled()) {
             logger.debug("[{}]: finishing early because the task was cancelled", task.getId());
             finishHim(null);
             return;
         }
         this.lastBatchSize = batchSize;
-        asyncResponse.done(worker.throttleWaitTime(thisBatchStartTimeNS, System.nanoTime(), batchSize));
+        this.totalBatchSizeInSingleScrollResponse += batchSize;
+
+        if (scrollRemainingHits.isEmpty()) {
+            int totalBatchSize = totalBatchSizeInSingleScrollResponse;
+            totalBatchSizeInSingleScrollResponse = 0;
+            asyncResponse.done(worker.throttleWaitTime(thisBatchStartTimeNS, System.nanoTime(), totalBatchSize));
+        } else {
+            final ScrollableHitSource.Response response = asyncResponse.response();
+            onScrollResponse(new ScrollableHitSource.AsyncResponse() {
+                @Override
+                public ScrollableHitSource.Response response() {
+                    return new ScrollableHitSource.Response(response.isTimedOut(),
+                        response.getFailures(),
+                        scrollRemainingHits.size(),
+                        scrollRemainingHits,
+                        response.getScrollId()
+                    );
+                }
+
+                @Override
+                public void done(TimeValue extraKeepAlive) {
+                    asyncResponse.done(extraKeepAlive);
```

Comment:
I think we could be building up a long chain of `AsyncResponse`'s, if every request fails to update anything? Bounded by the batch size so not too worrying, but would be nicer to have the delegation to the original `asyncResponse`.